Cleaned up auth server

WeaponMan opened the pull request:
1. Added root Cargo.toml with workspace, its just better to have open workspace then each project in seperate window
2. Changed collate on auth tables, because MariaDb doesnt have this collate 
3. Rewritten all the sqlx marcos to normal methods in wrath-auth-db, because marcos connect at build time to database, takes time and you cant build without existing database
3. Reworked packets into structs and traits, 
4. Reworked packet handlers, added wow_srp instead of manualy doing that work, I will have to create pr to it, because they expect plaintext password, but we got only sha1("username:password") in db

I do hope, you didnt abaddon this project, if not I will tackle the reconnects in auth server next.
If you need to slip into seperate PR let me know.

Victov commented:
Sorry for the late response, it's awesome that you contribute! I haven't abandoned the project officially, but I just have little time to work on it, currently. 

I didn't know that there was an existing srp crate. I like all your changes except the changes from query! to query. I prefer that my queries are checked during compile. I will gladly accept these changes, as they are almost all great improvements, if you could revert to using query! :) Thanks!

Victov commented on auth_server/src/main.rs:
```diff
@@ -39,23 +48,64 @@ async fn main() -> Result<()> {
+    let mut challenge_proof: Option<ServerChallengeProof> = None;
+    let mut log_on_proof: Option<ServerLogOnProof> = None;
 
+    //TODO(wmxd): keep some state that so the packets cant come out of order
```
The protocol uses TCP. Can packets even arrive out of order?

WeaponMan replied:
I mean that if you have rogue client that send logon proof before challenge, to that actual packets from network will be out of order.

WeaponMan replied:
It will be some kind of state machine later like:
challenge -> logon -> realm list
reconnect challenge -> (if we got the session_key for username) reconnect logon -> realm list

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
 
     pub async fn get_all_realms(&self) -> Result<Vec<DBRealm>> {
-        Ok(sqlx::query_as!(DBRealm, "SELECT * FROM realms").fetch_all(&self.connection_pool).await?)
+        Ok(sqlx::query_as::<_, DBRealm>("SELECT * FROM realms")
```
please revert to query_as!

WeaponMan replied:
reverted

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
 
     pub async fn set_realm_online_status(&self, realm_id: u32, online: bool) -> Result<()> {
-        sqlx::query!("UPDATE realms SET online = ? WHERE id = ?", online as u8, realm_id)
+        sqlx::query("UPDATE realms SET online = ? WHERE id = ?")
```
please revert to query!

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
 
     pub async fn set_realm_population(&self, realm_id: u32, population: f32) -> Result<()> {
-        sqlx::query!("UPDATE realms SET population = ? WHERE id = ?", population, realm_id)
+        sqlx::query("UPDATE realms SET population = ? WHERE id = ?")
```
please revert to macro variant query!

WeaponMan replied:
reverted

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
 
     pub async fn get_account_by_username(&self, username: &str) -> Result<DBAccount> {
-        let acc = sqlx::query_as!(DBAccount, "SELECT * FROM accounts WHERE username = ?", username)
+        let acc = sqlx::query_as::<_, DBAccount>("SELECT * FROM accounts WHERE username = ?")
```
please revert to macro variant query_as!

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
 
     pub async fn set_account_v_s(&self, account_id: u32, v: &str, s: &str) -> Result<()> {
-        sqlx::query!("UPDATE accounts SET v = ?, s = ? WHERE id = ?", v, s, account_id)
+        sqlx::query("UPDATE accounts SET v = ?, s = ? WHERE id = ?")
```
please revert to macro variant query!

WeaponMan replied:
reverted

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
 
     pub async fn set_account_sessionkey(&self, username: &str, session_key: &str) -> Result<()> {
-        sqlx::query!("UPDATE accounts SET sessionkey = ? WHERE username = ?;", session_key, username)
+        sqlx::query("UPDATE accounts SET sessionkey = ? WHERE username = ?;")
```
please revert to macro variant query!

WeaponMan replied:
reverted

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
 
     pub async fn get_account_data(&self, account_id: u32) -> Result<Vec<DBAccountData>> {
-        let acc_data = sqlx::query_as!(DBAccountData, "SELECT * FROM account_data WHERE account_id = ?", account_id)
+        let acc_data = sqlx::query_as::<_, DBAccountData>("SELECT * FROM account_data WHERE account_id = ?")
```
please revert to macro variant query_as!

WeaponMan replied:
reverted

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
-            "SELECT * FROM account_data WHERE account_id = ? AND data_type = ?",
-            account_id,
-            data_type
+        Ok(
```
please revert to macro variant query_as!

WeaponMan replied:
reverted

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
-        )
-        .execute(&self.connection_pool)
-        .await?;
+        sqlx::query("INSERT INTO account_data (account_id, data_type, time, decompressed_size, data) VALUES (?,?, 0, 0, NULL)")
```
please revert to macro variant query!

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
-        )
-        .execute(&self.connection_pool)
-        .await?;
+        sqlx::query("UPDATE account_data SET decompressed_size = ?, data = ?, time = ? WHERE account_id = ? AND data_type = ?")
```
please revert to macro variant query!

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
-        )
-        .fetch_one(&self.connection_pool)
-        .await;
+        let res = sqlx::query("SELECT num_characters FROM realm_characters WHERE account_id = ? AND realm_id = ?")
```
please revert to macro variant query!

Victov commented on databases/wrath-auth-db/src/lib.rs:
```diff
@@ -28,105 +30,111 @@ impl AuthDatabase {
 
     pub async fn set_num_characters_on_realm(&self, account_id: u32, realm_id: u32, num_characters: u8) -> Result<()> {
-        sqlx::query!("REPLACE INTO realm_characters VALUES (?, ?, ?)", account_id, realm_id, num_characters)
+        sqlx::query("REPLACE INTO realm_characters VALUES (?, ?, ?)")
```
please revert to macro variant query!

Victov reviewed (changes requested):
Some minor changes requested, mostly related to query macros

WeaponMan commented:
I dropped the query changes and rebased.

I had chat with the author of wow_srp:

- reconnect is insecure, I can implemented but will be good idea to put it behind some feature so it can be build without it and figure out for how long we will keep pair of session_key and username alive for reconnect be successfull if thats enabled
- api for already done `sha1 (username : password )` it is no acceptable, I will keep my fork alive for time being and in a future will be good idea not to generate  `v` and `s` in the auth server, but on actual user register/password change

ffvictor commented:
That's a good future improvement! I'm personally not very interested in improving the Auth Server, I just want to be able to login and develop the world server. But I love your enthusiasm to improve the auth server, so more pull requests are more than welcome!

Maybe we can modify the auth server to accept command line input, and we can add a `createaccount <user> <pass>` command that will take care of the `v` and `s` calculation and insert the user into the database. I'm not expecting that I'll create a full website for account creation. If you have another suggestion, I'm eager to hear it :)

Victov commented:
That's a great future improvement. The Auth server is not a huge priority for me, I just want to be able to login and develop the world server. But I love your enthusiasm to improve it, and future pull requests are more than welcome. 

Maybe we can modify the auth server to accept command line input, and create an `account create <user> <pass>` command that takes care of calculating `v` and `s` and then inserts into the database. It's nice and simple, which is perfect since I don't expect to make a full website for account management, since the project is soooo far from production-ready gameplay. 

If you have alternative suggestions, i'm eager to hear them :)

WeaponMan commented:
>That's a great future improvement. The Auth server is not a huge priority for me, I just want to be able to login and develop the world server. But I love your enthusiasm to improve it, and future pull requests are more than welcome.

I just wanted to start with something easy and get into it.

> Maybe we can modify the auth server to accept command line input, and create an account create <user> <pass> command that takes care of calculating v and s and then inserts into the database

Good idea, I will do that when I get some time over the weekend.

Victov commented:
@WeaponMan since merging these changes I can no longer log in on the realm with the test user. The world server rejects the session that the wow-srp creates, and the world server isn't updated yet to use wow-srp. How did you test this locally? Did you manage to go into character creation?

Perhaps you could do a pull request to update the world server to also use wow-srp?

WeaponMan commented:
It will be the hex format in the database, I will do fix for it.

WeaponMan commented:
Got fix for on my fork, cannot do rebase right now, power went out.
